File: q15.c
#define __STDC_FORMAT_MACROS
#include <inttypes.h>
#include <stdio.h>

int xx = 20;
int yy = 20;

uint64_t step(int x, int y, uint64_t total) {
    if (x == xx && y == yy)
        return total + 1;
    if (x < xx)
        total = step(x + 1, y, total);
    if (y < yy)
        total = step(x, y + 1, total);
    return total;
}

int main(int argc, char** argv) {
  printf("%" PRIu64 "\n", step(0, 0, 0));
}

File: q15.py
xx, yy = 4, 4


def step(x, y, total=0):
    if x < 13 and x == y:
        print '(%d, %d)' % (x, y)
    if x == xx and y == yy:
        return total + 1
    if x < xx:
        total = step(x + 1, y, total)
    if y < yy:
        total = step(x, y + 1, total)
    return total

print step(0, 0)
